compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeParamAnalyzer.java

```diff
@@ -306,6 +309,10 @@ private void findTypeParam(DiagnosticPos pos, BType expType, BType actualType, S
                     findTypeParamInMapForRecord(pos, (BMapType) expType, (BRecordType) actualType, env, resolvedTypes,
                                                 result);
                 }
+                if (actualType.tag == TypeTags.UNION) {
+                    findTypeParamInUnionForMap(pos, (BMapType) expType, (BUnionType) actualType, env, resolvedTypes,
+                                               result);
+                }
```

Comment:
The issue we are trying to solve is not only for lists and maps. I think you have to handle the union scenarios for pretty much each case here.

For example, for streams also it fails atm.
```ballerina
type Foo record {
    string id;
};

type Bar record {
    int id;
};


public function main() { 
    
    Foo[] fooList = [{id: "1234"}, {id: "5678"}];
        
    stream<Foo> fooStream = fooList.toStream();

    stream<Foo>|stream<Bar> fooBarStream = fooStream;

    record {|Foo|Bar value;|}|error? res = fooBarStream.next(); 
}
```

Reply:
Related to : https://github.com/ballerina-platform/ballerina-lang/issues/22570